py/bitbox02/bitbox02/util.py

```diff
@@ -39,3 +46,97 @@ def parse_xpub(xpub: str) -> common.XPub:
+            return
+        (privkey, pubkeys) = UserCache.deserialize(raw_cache)
+        self.app_static_privkey = privkey
+        self.device_static_pubkeys = pubkeys
+
+    def serialize(self) -> str:
+        """Serialize struct to string"""
+        pubkeys = [binascii.hexlify(x).decode("utf-8") for x in self.device_static_pubkeys]
+        privkey = None
+        if self.app_static_privkey is not None:
+            privkey = binascii.hexlify(self.app_static_privkey).decode("utf-8")
+        return json.dumps({"device_static_pubkeys": pubkeys, "app_static_privkey": privkey})
+
+    @staticmethod
+    def deserialize(raw: str) -> Tuple[Optional[bytes], List[bytes]]:
+        """Deserialize content from disk to struct"""
+        try:
+            data = json.loads(raw)
+            privkey = None
+            if data["app_static_privkey"] is not None:
+                privkey = binascii.unhexlify(data["app_static_privkey"])
+            pubkeys = [binascii.unhexlify(x) for x in data["device_static_pubkeys"]]
+            return (privkey, pubkeys)
+        except json.JSONDecodeError:
+            return (None, [])
+
+
+class NoiseConfigUserCache(bitbox_api_protocol.BitBoxNoiseConfig):
+    """A noise config that stores the keys in a file in XDG_CACHE_HOME or ~/.cache"""
+
+    def __init__(self, appid: str) -> None:
+        """
+        Args:
+            appid: A string that uniqely identifies your application. It will be used as the name
+            of the cache directory. Directory separators will create subdirectories, e.g.
+            "shift/test1".
+        """
+        self._cache_file_path = NoiseConfigUserCache._find_cache_file(appid)
+        super().__init__()
+
+    @staticmethod
+    def _find_cache_file(appid: str) -> Path:
+        cachedir_env = os.environ.get("XDG_CACHE_HOME", "")
+        if cachedir_env == "":
+            homedir = os.environ.get("HOME", "")
+            if homedir == "":
+                raise RuntimeError("Can't find cache dir")
+            cachedir = Path(homedir) / ".cache"
+        else:
+            cachedir = Path(cachedir_env)
+        return cachedir / appid / "bitbox02.dat"
+
+    def _read_cache(self) -> UserCache:
+        try:
+            with self._cache_file_path.open("r") as fileh:
+                return UserCache(fileh.read())
+        except FileNotFoundError:
+            return UserCache()
+
+    def _write_cache(self, data: UserCache) -> None:
+        self._cache_file_path.parent.mkdir(parents=True, exist_ok=True)
+        with self._cache_file_path.open("w") as fileh:
+            fileh.write(data.serialize())
+
+    def contains_device_static_pubkey(self, pubkey: bytes) -> bool:
+        data = self._read_cache()
+        if pubkey in data.device_static_pubkeys:
+            return True
+        return False
+
+    def add_device_static_pubkey(self, pubkey: bytes) -> None:
+        if not self.contains_device_static_pubkey(pubkey):
```

Comment:
What's with the empty case?

Reply:
What do you mean? A `UserCache` will always have the two members `device_static_pubkeys` and `app_static_privkey`.

Reply:
Yes, all good :+1: